### PR TITLE
Fixes #16 by resolving the correct node definition

### LINKED ISSUE #16: The context node is wrong sometimes
When the context node (under the cursor) is resolved, the returned ``XsdNode`` definition sometimes is wrong. This is because there can be different nodes with the same name in the schema, for example, if we look for the node ``inputs``, there are ``tool/inputs`` or  ``tool/configfiles/inputs`` since right now only a simple name search is used to determine the node, this can return the wrong node definition.
This can be fixed by trying to use the whole node path when searching for the node definition.

### PR DESCRIPTION
**Pull request by davelopez**
* Added a [function](https://github.com/davelopez/galaxy-tools-extension/compare/fix-contex-node?expand=1#diff-81db3ec65dc6328bb030ccf2859f5bfcR125-R144) to resolve the node definition (XSD) using the current path of node names.

* Fixed how the [node stack is calculated](https://github.com/davelopez/galaxy-tools-extension/compare/fix-contex-node?expand=1#diff-fc334655ded569ca2404b0b758a8d677R268-R296) when the XML is broken or incomplete.

* Added some tests for edge cases.

### CODE REVIEW

File: server/services/context.py [hunk 1]
```diff
@@ -264,11 +265,35 @@ def _try_recover_context_from_exception(self, start_offset: int) -> None:
...
 
     def _try_get_tag_context_at_line_position(self, target_offset):
-        tag_match = re.search(START_TAG_REGEX, self._context.document_line, re.DOTALL)
-        if tag_match and tag_match.start(TAG_GROUP) <= target_offset <= tag_match.end(TAG_GROUP):
-            self._context.token_type = ContextTokenType.TAG
-            self._context.token_name = tag_match.group(TAG_GROUP)
-            raise ContextFoundException()
+        tag_matches = re.finditer(START_TAG_REGEX, self._context.document_line, re.DOTALL)
+        if tag_matches:
+            for matchNum, match in enumerate(tag_matches, start=1):
```
Comment (bgruening) [inline server/services/context.py:270]: matchNum is not used and can be removed togeter with enumerate?

File: server/services/context.py [hunk 2]
```diff
@@ -264,11 +265,35 @@ def _try_recover_context_from_exception(self, start_offset: int) -> None:
...
+                if (
+                    target_offset > tag_end
+                    and tag not in self._context.node_stack
+                    and not self._is_tag_closed_before_offset(tag, target_offset)
+                ):
+                    self._context.node_stack.append(tag)
+
+    def _is_tag_closed_before_offset(self, tag: str, offset: int) -> bool:
+        match = re.search(f"</{tag}>", self._context.document_line)
+        if match:
```
Comment (bgruening) [inline server/services/context.py:289]: one if condition instead of two?

File: server/services/context.py [hunk 3]
```diff
@@ -264,11 +265,35 @@ def _try_recover_context_from_exception(self, start_offset: int) -> None:
...
+            if offset >= match.end(0):
+                return True
+        match = re.search(f"<{tag}[^>]*/>", self._context.document_line)
+        if match:
```
Comment (bgruening) [inline server/services/context.py:293]: smae here

File: server/services/context.py (hunk 1 above)
Comment (davelopez) [reply server/services/context.py:270]: Well spotted! Leftovers from the regex investigation :)

Comment (bgruening) [reply server/services/context.py:270]: I'm actually surprised that linting has not caught that.

Comment (davelopez) [reply server/services/context.py]: Yes, normally it cries, I am also surprised

**Comment by davelopez** [comment]
Thanks so much for the review! A good pair of eyes seems to be best than any linter :)

**Review by bgruening — APPROVED** [review]
